<br>

<img src="./image/Logo/logo_elia_group.png" width = 200>

<br>

# Advanced Manipulation
<br>

It is time to learn some advanced data manipulation. In the follwong section, you will learn how to access your data through looping. Looping is a super important method to, for instance, calculate new values based on your existing values. There are many different looping methods such as `iterrows()`, `apply()` or the vectorization method. The main difference between those is their speed. To simplify things, you will first focus on `apply()` and subsequently get to know vectorization. But don't worry. Vectorization sounds like hardcore math, but it isn't. 

<ins>You will  learn how to:</ins>
1. access and manipulate data in Series / DataFrames with **apply()**
2. Speed up the process by using **vectorization** methods

Let's get to it!

## Looping with apply()
<br>

A simple Dataframe consists of Series, which you know as columns. If you would now like to iterate and apply a function on your data, this would probably look a bit like this: 

`for row in range(len(df_short)): #  number of rows
    for col in range(len(df_short.columns)): # number of columns
        print(df_short.iloc[row, col])`
        
But there is a **way more easy way** than using loops! With `apply()` you can literally apply a function along a row or a column! First, let's have a look a Series before we move on to Dataframes:

### Accessing/Manipulating data in <ins>Series</ins> with **.apply()**

- `apply()` applies a function to each element in a Series e.g. calculate the string length of each value within a column
- check [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.apply.html) for more input

<img src = "./image/Icons/apply_column_example.png" width = 200>

Let's do an example on how to apply the `len()` function using `apply()`.
<br>

&#128161; About `len()`: This function is a Python built-in function. It returns the length of an object. For instance, it returns the number of items in a list or the lenght of a string.

In [3]:
import pandas as pd

In [80]:
physical_flow = pd.read_csv("./data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")
physical_flow.head()

Datetime Resolution code Asset Elia ID  \
0  2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
1  2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2  2022-05-30T23:15:00+02:00           PT15M        380_29   
3  2022-05-30T23:15:00+02:00           PT15M       380_111   
4  2022-05-30T23:15:00+02:00           PT15M        380_79   

                 Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   
3    GEZELLE-STEVIN 380.111  Internal Line         380.0    AC     121.828900   
4   AVELGEM-MASTAING 380.79        Tieline         380.0    AC       0.015625   

   Loading                          Asset geographic location  \
0     0.68  {"coordinates": [[4.28, 51.23], [4.3072839, 51...   
1     7.01  {"coordinates": [[3.6036937, 51.11457], [3.9, ...   
2    10.78  {"coordinates": [[4.181, 51.416], [4.24811, 51...   
3     3.72  {"coordinates": [[3.2011967, 51.2638468], [3.1...   
4     0.00  {"coordinates": [[3.4690219, 50.7824407], [3.4...   

                            geo_point_2d  
0                  51.2655739,4.29364195  
1  51.138284999999996,3.7518468499999997  
2                   51.39276245,4.214555  
3                    51.29485,3.19246695  
4                 50.59037085,3.43451095

Let's shorten the df for this example:

In [81]:
df_short = physical_flow.loc[:5, :"Asset type"]
df_short

Datetime Resolution code Asset Elia ID  \
0  2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
1  2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2  2022-05-30T23:15:00+02:00           PT15M        380_29   
3  2022-05-30T23:15:00+02:00           PT15M       380_111   
4  2022-05-30T23:15:00+02:00           PT15M        380_79   
5  2022-05-30T23:15:00+02:00           PT15M  380_33_BRUEG   

                  Asset name     Asset type  
0       DOEL-MERCATOR 380.51  Internal Line  
1      HORTA-MERCATOR 380.74  Internal Line  
2   RILLAND-ZANDVLIET 380.29        Tieline  
3     GEZELLE-STEVIN 380.111  Internal Line  
4    AVELGEM-MASTAING 380.79        Tieline  
5  BRUEGEL-COURCELLES 380.33  Internal Line

And create a Series: 

In [15]:
series_short = df_short["Asset Elia ID"]
series_short

0    380_51_LILLO
1    380_74_HORTA
2          380_29
3         380_111
4          380_79
5    380_33_BRUEG
Name: Asset Elia ID, dtype: object

Now let's apply the `len()` function using `apply()`:

In [17]:
id_lenght = series_short.apply(len)
id_lenght

0    12
1    12
2     6
3     7
4     6
5    12
Name: Asset Elia ID, dtype: int64

See, it returns the legth of the string (in this case the Asset Elia ID) for each row of a Series.

### Accessing/Manipulating data in a <ins>DataFrame</ins> with **.apply()**

- you can apply the `apply()` function to either axis of a dataframe, this means you can loop through rows and columns
    - axis = 0 for column 
    - axis = 1 for row.

**axis = 0:**

<img src = "./image/loop_column.png" width = 600>

**axis = 1:**

<img src = "./image/loop_row.png" width = 600>

So let's check out our Dataframe again and shorten it, so that it is better manageable for this example:

In [21]:
physical_flow.head()

Datetime Resolution code Asset Elia ID  \
0  2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
1  2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2  2022-05-30T23:15:00+02:00           PT15M        380_29   
3  2022-05-30T23:15:00+02:00           PT15M       380_111   
4  2022-05-30T23:15:00+02:00           PT15M        380_79   

                 Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   
3    GEZELLE-STEVIN 380.111  Internal Line         380.0    AC     121.828900   
4   AVELGEM-MASTAING 380.79        Tieline         380.0    AC       0.015625   

   Loading                          Asset geographic location  \
0     0.68  {"coordinates": [[4.28, 51.23], [4.3072839, 51...   
1     7.01  {"coordinates": [[3.6036937, 51.11457], [3.9, ...   
2    10.78  {"coordinates": [[4.181, 51.416], [4.24811, 51...   
3     3.72  {"coordinates": [[3.2011967, 51.2638468], [3.1...   
4     0.00  {"coordinates": [[3.4690219, 50.7824407], [3.4...   

                            geo_point_2d  
0                  51.2655739,4.29364195  
1  51.138284999999996,3.7518468499999997  
2                   51.39276245,4.214555  
3                    51.29485,3.19246695  
4                 50.59037085,3.43451095

In [23]:
ph_flow_short = physical_flow.loc[:10,"Resolution code":"Loading"]
ph_flow_short

Resolution code Asset Elia ID                    Asset name     Asset type  \
0            PT15M  380_51_LILLO          DOEL-MERCATOR 380.51  Internal Line   
1            PT15M  380_74_HORTA         HORTA-MERCATOR 380.74  Internal Line   
2            PT15M        380_29      RILLAND-ZANDVLIET 380.29        Tieline   
3            PT15M       380_111        GEZELLE-STEVIN 380.111  Internal Line   
4            PT15M        380_79       AVELGEM-MASTAING 380.79        Tieline   
5            PT15M  380_33_BRUEG     BRUEGEL-COURCELLES 380.33  Internal Line   
6            PT15M        380_26         DOEL-ZANDVLIET 380.26  Internal Line   
7            PT15M       380_109  GEZELLE-VAN MAERLANT 380.109  Internal Line   
8            PT15M       380_107  GEZELLE-VAN MAERLANT 380.107  Internal Line   
9            PT15M        380_10          ACHENE-GRAMME 380.10  Internal Line   
10           PT15M        380_28    MAASBRACHT-VAN EYCK 380.28        Tieline   

    Base Voltage AC/DC  Physical Flow  Loading  
0          380.0    AC      13.754076     0.68  
1          380.0    AC     238.395599     7.01  
2          380.0    AC     232.342749    10.78  
3          380.0    AC     121.828900     3.72  
4          380.0    AC       0.015625     0.00  
5          380.0    AC      24.655853     1.77  
6          380.0    AC     298.491598    18.90  
7          380.0    AC       0.030151     0.00  
8          380.0    AC       0.030151     0.00  
9          380.0    AC     440.964253    24.38  
10         380.0    AC      57.433527     3.64

First, let's write a function that returns **True** if the `Asset type` is **Internal Line** and **False** if it is **Tieline**. 

In [24]:
def asset_check(row):
    if row["Asset type"] == "Internal Line":
        return True
    else:
        return False

Now create a new column called `Asset Check` that applies this newly defined function:

In [30]:
ph_flow_short.loc[:, "Asset Check"] = ph_flow_short.apply(asset_check, axis = 1)

In [28]:
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Asset Check  
0         380.0    AC      13.754076     0.68         True  
1         380.0    AC     238.395599     7.01         True  
2         380.0    AC     232.342749    10.78        False  
3         380.0    AC     121.828900     3.72         True  
4         380.0    AC       0.015625     0.00        False

&#128515; Can you spot the newly column with the previously defined function applied? Amazing, well done!

### Exercise

Let's do another one. <br>
1. First, create a new function called `flow_check` that returns ">200 MW" if the `Physical Flow` value is bigger than 200 MW and "<200 MW" if it is less than 200 MW.
2. Create a new column called `Flow Size` that applies this function onto your dataframe

### Advanced Exercise
In your dataframe, the physical flow indicates how heavily the line is loaded relative to the maximum possible line loading.

1. Create a new column that returns the maximum possible loading.
        - For instance, if the physical flow is 13.75 MW and that equals 0.68% loading, then the max. loading is 2022 MW. 
3. Name the new column "Max. Loading"

## Additional functions: Map() / Applymap()
<br>

Since you now know how to loop through functions, let's check out some additional functions that really come in handy!

- `map()` is used to substitute each value with another value
- `applymap()` is used for element-wise operations across the whole DataFrame

So what does this mean...?

### Map

`map()` is a Series method. It allows us to map existing values of a Series to a different set of values. For instance, imagine you have a set of data with one column called "Gender" consisting of the string values "male, "female" and diverse. You know would like to translate those values, so that male = 0, female = 1, and diverse = 2. To do so, you can use `map()`! 
- is a Series Method
- allows to map an existing value of a Series to a different set of values
- e.g. translate male and female into 0 and 1

&#128526; Let's practice!

In the energy sector, you probably won't have to deal with gender data. Howver, `map()` still comes in handy. In the following example, you could e.g. replace `Internal Line` with 1 and `Tieline` with 2. 

In [72]:
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Asset Check Flow Size  \
0         380.0    AC      13.754076     0.68         True   <200 MW   
1         380.0    AC     238.395599     7.01         True   >200 MW   
2         380.0    AC     232.342749    10.78        False   >200 MW   
3         380.0    AC     121.828900     3.72         True   <200 MW   
4         380.0    AC       0.015625     0.00        False   <200 MW   

   Max. Loading  
0     20.226582  
1     34.007931  
2     21.553131  
3     32.749704  
4      0.000000

In [73]:
ph_flow_short.loc[:,"Asset Type Numeric"] = ph_flow_short["Asset type"].map({"Internal Line": 1, "Tieline": 2})

In [74]:
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Asset Check Flow Size  \
0         380.0    AC      13.754076     0.68         True   <200 MW   
1         380.0    AC     238.395599     7.01         True   >200 MW   
2         380.0    AC     232.342749    10.78        False   >200 MW   
3         380.0    AC     121.828900     3.72         True   <200 MW   
4         380.0    AC       0.015625     0.00        False   <200 MW   

   Max. Loading  Asset Type Numeric  
0     20.226582                   1  
1     34.007931                   1  
2     21.553131                   2  
3     32.749704                   1  
4      0.000000                   2

Does the Syntax look familiar to you? Right, the input of the `map()` function is based on key-value pairs. 

### Applymap

Another function that comes in handy is `applymap()`. It is a dataframe method that applies a function to **every element** of a dataframe. <br>

<img src = "./image/applymap_example.png" width = 600>
Let's check out an example: 

In [79]:
df_applymap = ph_flow_short.loc[:,["Base Voltage", "Physical Flow", "Loading"]].applymap(int)
df_applymap.head()

Base Voltage  Physical Flow  Loading
0           380             13        0
1           380            238        7
2           380            232       10
3           380            121        3
4           380              0        0

As you can see, with `applymap()` you can e.g. select all the columns with numeric values in your dataframe and change their data type from float to int. 

## Speed up the process - **Vectorization**

If you do simple data manipulation, then `apply()` just works fine. However, it is not the fastes way to access and manipulate your data. You can speed up the process by using the vectorization method. In order to use vectorization, you finally have to import numpy first. Numpy is an open source Python library that’s used frequently in science and engineering. If you work with numeric data, you will definitely learn to love numpy!

<ins>What are vectorization methods ?</ins>
- applying a manipulation to a whole array/(vector), instead of single values 
- you have been indirectly using vectorization when you used e.g. groupyby

<ins>Why should we use it?</ins>
- you can use it to avoid looping row by row over our dataset and save a lot of time


But first things first: 

In [ ]:
import numpy as np 

Now imagine, you would like to add a new column to your dataframe called `new column`. This new column simply combines your two columns `Asset Elia ID` and `Asset type`

In [83]:
df_short.head()

Datetime Resolution code Asset Elia ID  \
0  2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
1  2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2  2022-05-30T23:15:00+02:00           PT15M        380_29   
3  2022-05-30T23:15:00+02:00           PT15M       380_111   
4  2022-05-30T23:15:00+02:00           PT15M        380_79   

                 Asset name     Asset type  
0      DOEL-MERCATOR 380.51  Internal Line  
1     HORTA-MERCATOR 380.74  Internal Line  
2  RILLAND-ZANDVLIET 380.29        Tieline  
3    GEZELLE-STEVIN 380.111  Internal Line  
4   AVELGEM-MASTAING 380.79        Tieline

In [91]:
df_short.loc[:,"Asset"] = df_short["Asset Elia ID"] + df_short["Asset type"]

In [94]:
df_short

Datetime Resolution code Asset Elia ID  \
0  2022-05-30T23:15:00+02:00           PT15M  380_51_LILLO   
1  2022-05-30T23:15:00+02:00           PT15M  380_74_HORTA   
2  2022-05-30T23:15:00+02:00           PT15M        380_29   
3  2022-05-30T23:15:00+02:00           PT15M       380_111   
4  2022-05-30T23:15:00+02:00           PT15M        380_79   
5  2022-05-30T23:15:00+02:00           PT15M  380_33_BRUEG   

                  Asset name     Asset type                      Asset  
0       DOEL-MERCATOR 380.51  Internal Line  380_51_LILLOInternal Line  
1      HORTA-MERCATOR 380.74  Internal Line  380_74_HORTAInternal Line  
2   RILLAND-ZANDVLIET 380.29        Tieline              380_29Tieline  
3     GEZELLE-STEVIN 380.111  Internal Line       380_111Internal Line  
4    AVELGEM-MASTAING 380.79        Tieline              380_79Tieline  
5  BRUEGEL-COURCELLES 380.33  Internal Line  380_33_BRUEGInternal Line

So let's turn one of the `apply()` functions into a vectorization. <br>

In [97]:
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Asset Check Flow Size  \
0         380.0    AC      13.754076     0.68         True   <200 MW   
1         380.0    AC     238.395599     7.01         True   >200 MW   
2         380.0    AC     232.342749    10.78        False   >200 MW   
3         380.0    AC     121.828900     3.72         True   <200 MW   
4         380.0    AC       0.015625     0.00        False   <200 MW   

   Max. Loading  Asset Type Numeric  
0     20.226582                   1  
1     34.007931                   1  
2     21.553131                   2  
3     32.749704                   1  
4      0.000000                   2

In [95]:
def flow_size(row):
    if row["Physical Flow"] >= 200:
        return ">200 MW"
    else: 
        return "<200 MW"

The first guess would be, to pass the whole vector/(Series) (instead of single rows as before) into your function and then make the calculation. But...

In [99]:
def flow_size(series):
    print(series)
    if series >200:
        return ">200 MW"
    else:
        return "<200 MW"

ph_flow_short.loc[:,"Flow Size"] = flow_size(ph_flow_short["Physical Flow"])

0      13.754076
1     238.395599
2     232.342749
3     121.828900
4       0.015625
5      24.655853
6     298.491598
7       0.030151
8       0.030151
9     440.964253
10     57.433527
Name: Physical Flow, dtype: float64


ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

**This throws us a value error** because Python does not know how to tell if a whole column is greater than 25. This is where Numpy comes in handy.

## np.where()
<br>

First, let's check out `np.where()`. This method is like the "if statement" in Excel.

- Syntax: 
           np.where(
                conditional statement -> bool array,
                series/array/function()/scalar if True,
                series/array/function()/scalar if False
           )

Let's try it out: 

In [107]:
import numpy as np

ph_flow_short.loc[:, "Flow Size by vec"] = np.where(
    ph_flow_short["Physical Flow"]>200, # <-- condition
    ">200 MW", # <-- return if true
    "<200 MW" #<-- return if false
    )

ph_flow_short.head(3)

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Asset Check Flow Size  \
0         380.0    AC      13.754076     0.68         True   <200 MW   
1         380.0    AC     238.395599     7.01         True   >200 MW   
2         380.0    AC     232.342749    10.78        False   >200 MW   

   Max. Loading  Asset Type Numeric     Test Flow Size by vec  
0     20.226582                   1  <200 MW          <200 MW  
1     34.007931                   1  >200 MW          >200 MW  
2     21.553131                   2  >200 MW          >200 MW

## np.select()
<br>

If you have multiple conditions, you can simply use `np.select()`.
- Syntax: 
        conditions = [
            condition1
            condition2
            etc.
        ]

        choices = [
            value1
            value2
            etc.
        ]

df["new column"] = np.select(conditions, choices, default="NA")

Let's try out another example in regard to the size of the physical flow: 

In [115]:
conditions = [
    ph_flow_short["Physical Flow"] == 0, # first condition to test: if true return choice1, if false check next condition
    ph_flow_short["Physical Flow"] < 200, # second condition to test: if true return choice2, if false check next condition
    ph_flow_short["Physical Flow"] < 500 # third condition to test: if true return choice3, if false default value is returned
]

choices = [
    "None",  # choice1
    "Small", # choice2
    "Big",   # choice3
]

ph_flow_short.loc[:,"Flow_grouped"] = np.select(conditions, choices, default="Large") #  default value is the value if non of the conditions are true
print(ph_flow_short.Flow_grouped.value_counts())
ph_flow_short.tail(10)

Small    7
Big      4
Name: Flow_grouped, dtype: int64


Resolution code Asset Elia ID                    Asset name     Asset type  \
1            PT15M  380_74_HORTA         HORTA-MERCATOR 380.74  Internal Line   
2            PT15M        380_29      RILLAND-ZANDVLIET 380.29        Tieline   
3            PT15M       380_111        GEZELLE-STEVIN 380.111  Internal Line   
4            PT15M        380_79       AVELGEM-MASTAING 380.79        Tieline   
5            PT15M  380_33_BRUEG     BRUEGEL-COURCELLES 380.33  Internal Line   
6            PT15M        380_26         DOEL-ZANDVLIET 380.26  Internal Line   
7            PT15M       380_109  GEZELLE-VAN MAERLANT 380.109  Internal Line   
8            PT15M       380_107  GEZELLE-VAN MAERLANT 380.107  Internal Line   
9            PT15M        380_10          ACHENE-GRAMME 380.10  Internal Line   
10           PT15M        380_28    MAASBRACHT-VAN EYCK 380.28        Tieline   

    Base Voltage AC/DC  Physical Flow  Loading  Asset Check Flow Size  \
1          380.0    AC     238.395599     7.01         True   >200 MW   
2          380.0    AC     232.342749    10.78        False   >200 MW   
3          380.0    AC     121.828900     3.72         True   <200 MW   
4          380.0    AC       0.015625     0.00        False   <200 MW   
5          380.0    AC      24.655853     1.77         True   <200 MW   
6          380.0    AC     298.491598    18.90         True   >200 MW   
7          380.0    AC       0.030151     0.00         True   <200 MW   
8          380.0    AC       0.030151     0.00         True   <200 MW   
9          380.0    AC     440.964253    24.38         True   >200 MW   
10         380.0    AC      57.433527     3.64        False   <200 MW   

    Max. Loading  Asset Type Numeric     Test Flow Size by vec Flow_grouped  
1      34.007931                   1  >200 MW          >200 MW          Big  
2      21.553131                   2  >200 MW          >200 MW          Big  
3      32.749704                   1  <200 MW          <200 MW        Small  
4       0.000000                   2  <200 MW          <200 MW        Small  
5      13.929860                   1  <200 MW          <200 MW        Small  
6      15.793206                   1  >200 MW          >200 MW          Big  
7       0.000000                   1  <200 MW          <200 MW        Small  
8       0.000000                   1  <200 MW          <200 MW        Small  
9      18.087131                   1  >200 MW          >200 MW          Big  
10     15.778442                   2  <200 MW          <200 MW        Small

### Exercise 

This exercise is all about vectorization: 

- Check out the following dataframe
- Create a new column "Loading_grouped" in the DataFrame `flow_info` that groups the Loding (in %) using vectorization
- Loading less then 5 % is categorized as "small", less then 20 % as "medium" and more then 20 % as "large"
- print out the the last 10 rows of the updates dataframe

In [118]:
physical_flow = pd.read_csv("./data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")

flow_info = physical_flow.loc[:, "Asset name": "Loading"]
flow_info.head(3)

Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   

   Loading  
0     0.68  
1     7.01  
2    10.78

In [136]:
conditions = [
    flow_info["Loading"] < 5, # first condition to test: if true return choice1, if false check next condition
    flow_info["Loading"] <= 20, # second condition to test: if true return choice2, if false check next condition
    flow_info["Loading"] > 20 # third condition to test: if true return choice3, if false default value is returned
]

choices = [
    "small",  # choice1
    "medium", # choice2
    "large",   # choice3
]

flow_info.loc[:,"Loading_grouped"] = np.select(conditions, choices) #  default value is the value if non of the conditions are true

In [122]:
print(flow_info.Loading_grouped.value_counts())
flow_info.tail(10)

medium    2170
large     1658
small     1248
Name: Loading_grouped, dtype: int64


Asset name     Asset type  Base Voltage AC/DC  \
5066  MAASBRACHT-VAN EYCK 380.28        Tieline         380.0    AC   
5067         GRAMME-LIXHE 380.11  Internal Line         380.0    AC   
5068        DOEL-MERCATOR 380.52  Internal Line         380.0    AC   
5069  HORTA-VAN MAERLANT 380.104  Internal Line         380.0    AC   
5070        DOEL-MERCATOR 380.51  Internal Line         380.0    AC   
5071    COURCELLES-GRAMME 380.31  Internal Line         380.0    AC   
5072       HORTA-MERCATOR 380.74  Internal Line         380.0    AC   
5073     BRUEGEL-MERCATOR 380.36  Internal Line         380.0    AC   
5074  HORTA-VAN MAERLANT 380.103  Internal Line         380.0    AC   
5075   BRUEGEL-COURCELLES 380.34  Internal Line         380.0    AC   

      Physical Flow  Loading Loading_grouped  
5066      64.885999     4.11           small  
5067     254.106699    14.87          medium  
5068     550.735256    37.37           large  
5069     248.927759     7.20          medium  
5070       2.042895     0.10           small  
5071     154.208288     8.47          medium  
5072      95.991292     3.32           small  
5073     748.364497    47.38           large  
5074     207.980390     6.19          medium  
5075      42.436825     2.69           small

### Advanced Exercise 

In the follwing, you will calculate the maximum load again. Similar to the exercise from above. However, this time you will use vectorization again.

- Work with the dataframe `flow_info`
- Create a new column called "Max_load" that shows the maximum possible load
        - For instance, if the physical flow is 13.75 MW and that equals 0.68% loading, then the max. loading is 2022 MW
- Hint: Pay attention to zero values

In [138]:
flow_info.loc[:, "Max_load"] = np.where(
    (flow_info["Physical Flow"] > 0) & (flow_info["Loading"] > 0), # <-- condition
    flow_info["Physical Flow"] / flow_info["Loading"] * 100, # <-- return if true
    0 #<-- return if false
    )
flow_info.head(3)

Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   

   Loading Loading_grouped     Max_load  
0     0.68           small  2022.658205  
1     7.01          medium  3400.793142  
2    10.78          medium  2155.313073

<br>

## Recap, Tips & Takeaways &#128161;

<br>

<div class="alert alert-block alert-success">

**Let's recap what you have learned in this section:**

- You can use `apply()` to access and manipulate Series and Dataframes
- In a Dataframe, you can loop through rows and columns, you just need to specify the axis: 
    - axis = 0 for column 
    - axis = 1 for row
- You can speed up the process by using vectorization
    - Here, you need to import numpy as np
    - `np.where()` is basically like an if statement in Excel
    - `np.select()` can be used to define multiple if statements with its conditions and choices 
 
</div>